### DISCUSSION TITLE
ESPnet installation using WSL-2 (running Ubuntu 20.04)

### DISCUSSION THREAD

**Discussion by Bereket-Desbele** [espnet/espnet#4909] [Q&A] [answered]
Hello! Has anyone ever tried installing ESPNET on windows 11 using WSL-2? If so, is it possible? I recently purchased an MSI laptop but unfortunatly i learned that it doesnot have linux drivers. So i though of using WSL to circumvent this issue. Now am looking for any advice i can get. Many thanks inadvance.

**Comment by Bereket-Desbele** [comment]
I've nearly completed the setup process successfuly. Unfortunately the training terminates on stage 8 of egs2 recepies. The logs aren't of much help as well.

**Comment by sw005320** [comment, marked as answer]
Can you display the log file?
https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/asr1/asr.sh#L982 is just for the calculation of the LM perplexity.
I don't know why, but you could actually skip it.

**Reply by Bereket-Desbele** [reply]
I have skipped stage-8 and then, another error is fired during asr-training stage. Attached are the log files of LM-training stage, lm perplexity stage and asr-training stage. & i can't figureout why, but the LM-perplexity is an empty log file. Probably because the training haults abrapty. In my opionion the other log files ought to help figureout the problem though. 

[lm-train.log](https://github.com/espnet/espnet/files/10682792/lm-train.log)
[asr-train.log](https://github.com/espnet/espnet/files/10682793/asr-train.log)
[lm_calc_perplexity.log](https://github.com/espnet/espnet/files/10682794/lm_calc_perplexity.log)

**Reply by Bereket-Desbele** [reply]
Could maybe it have ranout of memory? Similar errors pointed out by others say so. 

PS. The logs are for an4 recepie on espnet2. It's currently running on a laptop with intel-i7 12th Generation, 16GB DDDR4 RAM and Nvidia-GeForce RTX-3070 laptop GPU (8GB GDDR6) memory. This is all ofcourse running on windows-11 22H2 under a WSL-2 (windows subsystem for linux) setup with ubuntu-20.04.

**Reply by sw005320** [reply]
```
RuntimeError: Unable to find a valid cuDNN algorithm to run convolution
```

It seems that the issue comes from your GPU/cuDNN environments.
Not from espnet.
I think it is better to find a solution in PyTorch.

**Reply by Bereket-Desbele** [reply]
okay, i will have a look at it. but for now, could you atleast please help me change the training batch-size so i could try to see if it could fix the issue. Where can i find the training batch-size value??

**Reply by sw005320** [reply]
See `https://github.com/espnet/espnet/blob/master/egs2/an4/asr1/conf/train_asr_transformer.yaml#L2` and also related document, e.g., https://espnet.github.io/espnet/espnet2_training_option.html#change-mini-batch-type

**Reply by Bereket-Desbele** [reply]
Thanks. But I think the error msg, "RuntimeError: Unable to find a valid cuDNN algorithm to run convolution" is misleading. I just lowered the value of batch-size from 64 to 32 and the training is going smooth now. As for the LM-perplexity calc, i still cant find any solutions. Can you suggest any params i should tweak to try and make that work.

**Reply by sw005320** [reply]
OK, sounds good.

> As for the LM-perplexity calc, i still cant find any solutions. Can you suggest any params i should tweak to try and make that work.

I have no ideas.
Your log cannot be displayed in my environment.
It is probably because yur LM training is not finished succesfully?

**Reply by Bereket-Desbele** [reply]
[train.log](https://github.com/espnet/espnet/files/10685594/train.log)
here try this. It is the language model's training log. No errors on it though. Perhaps there is also another param i have tweak, to successfuly complete stage 8  (LM-perplexity calc)

**Reply by Bereket-Desbele** [reply]
Hi again. It's finished asr training and has haulted on decoding stage with similar error msg as for stage 8. here is the log file from stage 8. 
[lm_calc_perplexity.log](https://github.com/espnet/espnet/files/10686638/lm_calc_perplexity.log)

As for the decoding section, here is the output error on screen 

Traceback (most recent call last):
  File "/home/bereket/espnet-master/tools/anaconda/envs/espnet/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/bereket/espnet-master/tools/anaconda/envs/espnet/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/bereket/espnet/espnet2/bin/asr_inference.py", line 902, in <module>
    main()
  File "/home/bereket/espnet/espnet2/bin/asr_inference.py", line 898, in main
    inference(**kwargs)
  File "/home/bereket/espnet/espnet2/bin/asr_inference.py", line 611, in inference
    speech2text = Speech2Text.from_pretrained(
  File "/home/bereket/espnet/espnet2/bin/asr_inference.py", line 515, in from_pretrained
    return Speech2Text(**kwargs)
  File "/home/bereket/espnet/espnet2/bin/asr_inference.py", line 149, in __init__
    lm, lm_train_args = LMTask.build_model_from_file(
  File "/home/bereket/espnet/espnet2/tasks/abs_task.py", line 1866, in build_model_from_file
    model.load_state_dict(torch.load(model_file, map_location=device))
  File "/home/bereket/espnet-master/tools/anaconda/envs/espnet/lib/python3.8/site-packages/torch/serialization.py", line 795, in load
    return _legacy_load(opened_file, map_location, pickle_module, **pickle_load_args)
  File "/home/bereket/espnet-master/tools/anaconda/envs/espnet/lib/python3.8/site-packages/torch/serialization.py", line 1002, in _legacy_load
    magic_number = pickle_module.load(f, **pickle_load_args)
EOFError: Ran out of input
# Accounting: time=6 threads=1
# Ended (code 1) at Wed Feb  8 17:31:04 MSK 2023, elapsed time 6 seconds

**Reply by sw005320** [reply]
Does the model actually exist?
You can write a simple code to test the model load function.

**Reply by Bereket-Desbele** [reply]
Decoding with CPU works, as for with GPU, it throws the following error.
[asr_inference.1.log](https://github.com/espnet/espnet/files/10730308/asr_inference.1.log)


What could be the solution to tjis

**Reply by sw005320** [reply]
I usually do not test the GPU inference.
@pyf98, do you know what's happening?

**Reply by Bereket-Desbele** [reply]
I also have noticed that the GPU utilization is not efficient. I came across an article benchmarking a bare metal & a WSL2 Linux OS GPU utilization, it said more work should be given to the GPU so the performance difference gets negligible (minimal overheard) due to the asynchronous API calls. 

What could be done to improve the inner workings of espnet as a whole to improve its efficiency when run using WSL2. 


Here is the article https://developer.nvidia.com/blog/leveling-up-cuda-performance-on-wsl2-with-new-enhancements/

**Reply by pyf98** [reply]
About the GPU inference error in the log file, it seems that some tensors are not in the same device. I think you can check which tensors are not properly moved to CUDA.

GPU decoding usually works well. I am not sure about the cause here.

**Comment by pyf98** [comment]
I also encountered the device mismatch error when using GPU decoding. I guess something was modified recently. I'll try to fix it.

**Reply by Bereket-Desbele** [reply]
Glad to hear. Let me know when you do.

**Reply by sw005320** [reply]
It is already merged. Can you try it with the latest master?
This is a bug, actually, and thanks for your report.
https://github.com/espnet/espnet/pull/4941

**Reply by Bereket-Desbele** [reply]
I'm glad i was of any help. & Yeah it's working fine now. Thank you both.

**Reply by Bereket-Desbele** [reply]
I'm glad i was of any help. & Yeah it's working fine now. Thank you both.

**Reply by sw005320** [reply]
Did you need any modifications except for the GPU memory issue above?
It would be very useful for the other people as well.

**Reply by Bereket-Desbele** [reply]
Attached is the instructions on Setting Up / Installing ESPNET on windows 11 using WSL-2 with GPU support. Along with possible errors one might face & their solutions. In my opinion this  would've been very helpful if it was included on the espnet installation docs. 

[Instructions.txt](https://github.com/espnet/espnet/files/10780845/Instructions.txt)

PS. My apologies for the delay. Internet access is huge problem from where i am from (Eritrea). 
Once again thanks for the help.

**Comment by sw005320** [comment]
Thanks!
I will add this information to the [installation page](https://github.com/espnet/espnet/edit/master/doc/installation.md).
https://github.com/espnet/espnet/pull/4949
